### ISSUE TITLE
習題 7：請用 micrograd 梯度引擎重做習題 1 的函數優化

### ISSUE THREAD

**Issue by ccckmit** [ccc113b/_ml#7]
函數

$$
f = x^2 + y^2 + z^2 - 2x - 4y - 6z + 8
$$

找最低點

參考

1. [02-microGrad](https://github.com/ccc113b/py2cs/tree/master/02-%E6%A9%9F%E5%99%A8%E5%AD%B8%E7%BF%92/01-%E5%84%AA%E5%8C%96%E7%AE%97%E6%B3%95/02-%E6%A2%AF%E5%BA%A6%E5%84%AA%E5%8C%96/04-%E5%8F%8D%E5%82%B3%E9%81%9E%E7%AE%97%E6%B3%95/02-microGrad) : [ex2.py](https://github.com/ccc113b/py2cs/blob/master/02-%E6%A9%9F%E5%99%A8%E5%AD%B8%E7%BF%92/01-%E5%84%AA%E5%8C%96%E7%AE%97%E6%B3%95/02-%E6%A2%AF%E5%BA%A6%E5%84%AA%E5%8C%96/04-%E5%8F%8D%E5%82%B3%E9%81%9E%E7%AE%97%E6%B3%95/02-microGrad/ex2.py)
2. https://github.com/karpathy/micrograd/
3. 數值版的梯度下降法 [gd.py](https://github.com/ccc113b/py2cs/blob/master/02-%E6%A9%9F%E5%99%A8%E5%AD%B8%E7%BF%92/01-%E5%84%AA%E5%8C%96%E7%AE%97%E6%B3%95/02-%E6%A2%AF%E5%BA%A6%E5%84%AA%E5%8C%96/03-%E6%A2%AF%E5%BA%A6%E4%B8%8B%E9%99%8D%E6%B3%95/gd.py) , [gdArray](https://github.com/ccc113b/py2cs/blob/master/02-%E6%A9%9F%E5%99%A8%E5%AD%B8%E7%BF%92/01-%E5%84%AA%E5%8C%96%E7%AE%97%E6%B3%95/02-%E6%A2%AF%E5%BA%A6%E5%84%AA%E5%8C%96/03-%E6%A2%AF%E5%BA%A6%E4%B8%8B%E9%99%8D%E6%B3%95/gdArray.py)

**Comment by liu99002** [comment]
資工三 111110506 劉銓旭
https://github.com/liu99002/_ml/blob/main/hw/hw7

**Comment by yc0123** [comment]
資工三 111110507 陳穎辰
https://github.com/yc0123/_ml/blob/main/hw7/hw7.py

**Comment by Erkmrcl17** [comment]
資工二 111210559 李麗恩
[HW7](https://github.com/Erkmrcl17/_ml/blob/main/HW6/micrograd/HW7)

**Comment by KiuBios** [comment]
資工三 111110519 胡劭家
https://github.com/KiuBios/_ml/tree/main/hw7

**Comment by jm123u** [comment]
資工四 111010577 史煜疏
https://github.com/jm123u/_ml/blob/main/hw7/micchange.py

**Comment by ChadChenYu** [comment]
#### 資工二 111210518 陳辰鈺
[由GPT-4o生成](https://github.com/ChadChenYu/_ml/blob/main/HW7micrograd.py)
[對話連結](https://chatgpt.com/share/67cf0bbf-fb08-8003-b592-30b46881d36a)

**Comment by Jonathan-es** [comment]
### 資工二 111210554 陳家盛
[HW7](https://github.com/Jonathan-es/_ml/blob/main/hw7/hw7.py)

**Comment by ruoxin030704** [comment]
資工四 111010576 王若馨
https://github.com/ruoxin030704/_ml/blob/main/hw7/hw7.py

**Comment by yang930208** [comment]
資工三 111110529 楊智鈞
[由ChatGPT生成](https://github.com/yang930208/_ml/blob/main/%E4%BD%9C%E6%A5%AD7/micrograd.py)
[對話連結](https://chatgpt.com/share/681c1910-4828-8008-97b6-cd465832fdd9)

**Comment by lkhagvaa2444** [comment]
資工三 111110544 李荷葦
https://github.com/lkhagvaa2444/_ml/tree/main/hw7

**Comment by 123456789203** [comment]
電機四乙 111012212 林子安
import random
from micrograd.engine import Value

def function(x, y, z):
    return x**2 + y**2 + z**2 - 2*x - 4*y - 6*z + 8

def gradient_descent(learning_rate=0.1, max_iterations=100, tolerance=1e-5):
    x = Value(random.uniform(-5, 5))
    y = Value(random.uniform(-5, 5))
    z = Value(random.uniform(-5, 5))

    for iteration in range(max_iterations):
        loss = function(x, y, z)
        
        x.grad = 0
        y.grad = 0
        z.grad = 0
        
        loss.backward()
        
        x.data -= learning_rate * x.grad
        y.data -= learning_rate * y.grad
        z.data -= learning_rate * z.grad
        
        if abs(loss.data) < tolerance:
            break

    return x.data, y.data, z.data, loss.data

x_opt, y_opt, z_opt, f_opt = gradient_descent()

print(f"x={x_opt:.3f}, y={y_opt:.3f}, z={z_opt:.3f}, f(x,y,z)={f_opt:.3f}")

**Comment by Steventanardi** [comment]
資工三 111110542 林源茂

https://github.com/Steventanardi/_ml/tree/main/Homework7

**Comment by wrr606** [comment]
資工三 111110517 范揚玄
https://github.com/wrr606/_ml/tree/main/hw/7/test.py

**Comment by henk3520** [comment]
資工四 111010573 古莛禾
https://github.com/henk3520/_ml/blob/main/hw7/hw7.py

**Comment by Yangforrok** [comment]
資工三 111110514 羊柏宇
https://github.com/Yangforrok/_ml/tree/main/07

**Comment by ouhanchen** [comment]
資工三 111110515 歐漢宸
https://github.com/ouhanchen/_ml/tree/main/7

**Comment by Gavinhsu0512** [comment]
資工三 111110501 許元承
https://github.com/Gavinhsu0512/_ml/tree/main/homework7

**Comment by Huang9315** [comment]
資工三黃志仁https://github.com/Huang9315/_ml/blob/main/hw7/micrograd.py

**Comment by Candy96551** [comment]
資工三 111110502 蔡雨軒
https://github.com/Candy96551/113_ml/blob/main/hw7/hw7.py

**Comment by sealbrother** [comment]
資工三 111110556 蔡杰叡
https://github.com/sealbrother/_ml/blob/main/hw7.py

**Comment by o5xNo** [comment]
資工三 111110508 吳丞恩
https://github.com/o5xNo/_ml/tree/main/homework/7.py

**Comment by karta9630** [comment]
資工三 111110503 戴嘉豪
https://github.com/karta9630/_ml/tree/main/HW7

**Comment by peterwang0329** [comment]
資工二 111210513 汪章貴
[https://github.com/peterwang0329/_ml/blob/main/hw6-7_microGrad/hw7_HillClimbingToMicrograd.py](https://github.com/peterwang0329/_ml/blob/main/hw6-7_microGrad/hw7_HillClimbingToMicrograd.py)

**Comment by chen02648** [comment]
資工四 111010539 陳文茜
https://github.com/chen02648/_ml/blob/main/hw7.py

**Comment by VinsOrl** [comment]
資工二 111210556 林新興
https://github.com/VinsOrl/_ml/tree/main/HW7

**Comment by patrick0516** [comment]
資工四 111010575 周辰恩
https://github.com/patrick0516/_ml/tree/main/hw7

**Comment by GFVictorique** [comment]
資工三 111010521 薛翊呈
https://github.com/GFVictorique/_ml/tree/main/hw7

**Comment by gakuplusq** [comment]
[資工四 黃劭騏 110910530](https://github.com/gakuplusq/_ml/wiki/hw7)

**Comment by XaioYue** [comment]
資工三 111110555 董翔宇
https://github.com/XaioYue/_ml/tree/main/7

**Comment by Remilia517** [comment]
資工四 111010532 李昱宏
[https://github.com/Remilia517/_ml/tree/main/hw/hw7](https://github.com/Remilia517/_ml/tree/main/hw/hw7)

**Comment by cindycallista** [comment]
楊莉思 資工二 111210555
https://github.com/cindycallista/_ml/blob/main/HW7.py

**Comment by sam00002135** [comment]
資工四 111010574 白和鑫
https://github.com/sam00002135/ML/blob/main/hm07/hm07.py

**Comment by mienweng** [comment]
資工三111110523 翁敏恩
https://github.com/mienweng/_ml/blob/main/homework/h7/h7.py

**Comment by ding1226** [comment]
資工三 111110546 丁虹羽
https://github.com/ding1226/_ml/blob/main/hw/hw7

**Comment by CJI3751** [comment]
資工四 111010564  潘敏驛
https://github.com/CJI3751/_ml/tree/main/homework/07

**Comment by HuangAlvin** [comment]
資工四 111010526 黃俊皓
https://github.com/HuangAlvin/_ml/blob/main/07.py

**Comment by yakumocirno** [comment]
資工三 111110551 劉品佑
https://github.com/yakumocirno/_ml/tree/main/hw7

**Comment by ghost** [comment]
資工三 111110525 林思伶
https://github.com/Siling5304/_ml/blob/main/hw7/hw7.py

**Comment by mimmm4308** [comment]
資工四111010525林品翔
https://github.com/mimmm4308/_ml/tree/main/final/07

**Comment by pado35** [comment]
資工三 111110511 李明洋
https://github.com/pado35/_ml/tree/main/7

**Comment by 921006vicky** [comment]
資工三 111110505 劉思妤https://github.com/921006vicky/_ml/tree/main/hw/07

**Comment by Kruez0** [comment]
資工三 111110541 劉佩芬
https://github.com/Kruez0/_ml/tree/main/w7

**Comment by Logan910115** [comment]
資工四 110910560 羅德耕https://github.com/Logan910115/_ml/blob/main/HW7.py

**Comment by Steventanardi** [comment]
資工四 111110542 林源茂 
https://github.com/Steventanardi/_cm/tree/main/Homework7
